cpp/src/progress_thread.cpp

```diff
@@ -84,36 +84,41 @@ void ProgressThread::remove_function(FunctionID function_id) {
         "Function was not registered with this ProgressThread"
     );
 
-    std::unique_lock lock(mutex_);
-    auto it = functions_.find(function_id.function_index);
-    RAPIDSMPF_EXPECTS(
-        it != functions_.end(), "Function not registered or already removed"
-    );
-
-    // Wait for the function to complete.
-    // iterator it can get invalidated, if some other thread erases a function. So, query
-    // functions_ instead
-    cv_.wait(lock, [&, f_idx = it->first]() { return functions_.at(f_idx).is_done; });
+    auto function_idx = function_id.function_index;
 
-    // Waiting done. Now, mutex_ is locked again
-    functions_.erase(function_id.function_index);
-
-    if (functions_.empty()) {
-        thread_.pause();
+    // First check if it's already completed
+    if (completion_tracker_.is_completed(function_idx)) {
+        completion_tracker_.remove(function_idx);
+        return;
     }
+
+    // Wait for completion
+    completion_tracker_.wait_for_completion(function_idx);
+    completion_tracker_.remove(function_idx);
 }
 
 void ProgressThread::event_loop() {
     auto const t0_event_loop = Clock::now();
+    std::vector<FunctionIndex> completed_functions;
     {
         std::lock_guard<std::mutex> lock(mutex_);
         for (auto& [id, function] : functions_) {
             function();
+            if (function.is_done) {
+                completed_functions.push_back(id);
+                completion_tracker_.mark_completed(id);
+            }
+        }
```

Comment:
@pentschev do you think we could traverse the `functions_` map without holding the lock? I think that is the bottleneck in the current impl. And suspect that's why @madsbk 's [PR of adding a small sleep](https://github.com/rapidsai/rapidsmpf/pull/253) helped.

Reply:
Yes, that is in essence what you've implemented as well, done in https://github.com/rapidsai/rapidsmpf/pull/265/commits/595647568e16dbac46c6ec4ef7112bf05e68a31a . (Again, not cleaned up for any resources that are now redundant)

Reply:
Let me go through this